src/main/java/io/netty/incubator/codec/http3/Http3FrameToHttpObjectCodec.java

```diff
@@ -86,31 +91,33 @@ protected void channelRead(ChannelHandlerContext ctx, Http3HeadersFrame frame, b
             return;
         }
 
-        if (isLast) {
-            if (headers.method() == null && status == null) {
-                LastHttpContent last = new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER, validateHeaders);
-                HttpConversionUtil.addHttp3ToHttpHeaders(id, headers, last.trailingHeaders(),
-                        HttpVersion.HTTP_1_1, true, true);
-                ctx.fireChannelRead(last);
-            } else {
-                FullHttpMessage full = newFullMessage(id, headers, ctx.alloc());
-                ctx.fireChannelRead(full);
-            }
+        if (headers.method() == null && status == null) {
+            // Must be trailers!
+            LastHttpContent last = new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER, validateHeaders);
+            HttpConversionUtil.addHttp3ToHttpHeaders(id, headers, last.trailingHeaders(),
+                    HttpVersion.HTTP_1_1, true, true);
+            inboundTranslationInProgress = false;
+            ctx.fireChannelRead(last);
         } else {
             HttpMessage req = newMessage(id, headers);
             if (!HttpUtil.isContentLengthSet(req)) {
                 req.headers().add(HttpHeaderNames.TRANSFER_ENCODING, HttpHeaderValues.CHUNKED);
             }
+            inboundTranslationInProgress = true;
             ctx.fireChannelRead(req);
         }
     }
 
     @Override
-    protected void channelRead(ChannelHandlerContext ctx, Http3DataFrame frame, boolean isLast) throws Exception {
-        if (isLast) {
-            ctx.fireChannelRead(new DefaultLastHttpContent(frame.content()));
-        } else {
-            ctx.fireChannelRead(new DefaultHttpContent(frame.content()));
+    protected void channelRead(ChannelHandlerContext ctx, Http3DataFrame frame) throws Exception {
+        inboundTranslationInProgress = true;
```

Comment:
can this be hit without a Http3HeadersFrame that already set this flag?

Reply:
this should never happen as a header frame must be sent first (which is validated by the frame decoder etc)